File: packages/cli/src/commands/memory.ts
import { Command } from "commander";
import chalk from "chalk";
import ora from "ora";
import { createApiClient } from "../lib/api-client.js";
import { success, info, keyValue } from "../lib/output.js";
import {
  handleError,
  confirmAction,
  validateStringLength,
  validateContentSize,
  validateEnum,
  validateUUID,
  validatePositiveInt,
} from "../lib/shared.js";

const VALID_TIERS = ["core", "active", "archive"];
const VALID_CATEGORIES = [
  "preference",
  "pattern",
  "decision",
  "correction",
  "knowledge",
  "general",
];

/* ─── remb memory ─── */

export const memoryCommand = new Command("memory")
  .description("Manage AI memories — add, list, update, delete, and promote")
  .addHelpText(
    "after",
    `\nExamples:\n  $ remb memory add -t "Auth pattern" -c "Uses JWT httpOnly cookies"\n  $ remb memory list --tier core\n  $ remb memory update <id> -c "Updated content"\n  $ remb memory delete <id>`,
  )
  .addCommand(memoryAddCommand())
  .addCommand(memoryListCommand())
  .addCommand(memoryUpdateCommand())
  .addCommand(memoryDeleteCommand())
  .addCommand(memoryPromoteCommand());

/* ─── remb memory add ─── */

function memoryAddCommand() {
  return new Command("add")
    .description("Create a new memory")
    .requiredOption("-t, --title <title>", "Memory title")
    .requiredOption("-c, --content <content>", "Memory content")
    .option("--tier <tier>", "Memory tier: core, active, archive", "active")
    .option("--category <category>", "Category: preference, pattern, decision, correction, knowledge, general", "general")
    .option("--tags <tags>", "Comma-separated tags")
    .option("-p, --project <slug>", "Project slug")
    .action(async (opts) => {
      validateStringLength(opts.title, "Title", 200);
      validateContentSize(opts.content, 50);
      validateEnum(opts.tier, "tier", VALID_TIERS);
      validateEnum(opts.category, "category", VALID_CATEGORIES);

      const spinner = ora("Creating memory...").start();
      try {
        const client = createApiClient();
        const result = await client.createMemory({
          title: opts.title,
          content: opts.content,
          tier: opts.tier,
          category: opts.category,
          tags: opts.tags ? opts.tags.split(",").map((t: string) => t.trim()) : undefined,
          projectSlug: opts.project,
        });
        spinner.stop();
        success(`Memory created`);
        keyValue("ID", result.id);
        keyValue("Tier", result.tier);
        keyValue("Category", result.category);
        keyValue("Tokens", String(result.token_count));
      } catch (err) {
        spinner.stop();
        handleError(err);
      }
    });
}

/* ─── remb memory list ─── */

function memoryListCommand() {
  return new Command("list")
    .alias("ls")
    .description("List memories — shows project-scoped and global memories")
    .option("--tier <tier>", "Filter by tier: core, active, archive")
    .option("--category <category>", "Filter by category")
    .option("-s, --search <query>", "Semantic/text search")
    .option("-p, --project <slug>", "Show memories for a project (includes global memories too)")
    .option("--global", "Show only global memories (no project scope)")
    .option("-l, --limit <n>", "Max results", "20")
    .option("--format <format>", "Output format: table, json, markdown", "table")
    .addHelpText(
      "after",
      `\nMemory scopes:\n  [project]  Scoped to a specific project — relevant to that codebase\n  [global]   No project scope — apply across all projects\n\nExamples:\n  $ remb memory list                          # All memories\n  $ remb memory list --project my-app         # Project + global memories\n  $ remb memory list --global                 # Only global memories\n  $ remb memory list --tier core -s "auth"    # Search core tier`,
    )
    .action(async (opts) => {
      const limit = parseInt(opts.limit, 10) || 20;
      validatePositiveInt(limit, "Limit", 200);
      if (opts.tier) validateEnum(opts.tier, "tier", VALID_TIERS);
      if (opts.category) validateEnum(opts.category, "category", VALID_CATEGORIES);

      const spinner = ora("Fetching memories...").start();
      try {
        const client = createApiClient();
        const listParams: Parameters<ReturnType<typeof createApiClient>["listMemories"]>[0] = {
          tier: opts.tier,
          category: opts.category,
          search: opts.search,
          limit,
        };
        // --global: no project filter, then filter client-side to null project_id
        // --project: pass project slug (API returns project + global)
        if (!opts.global && opts.project) listParams.project = opts.project;
        const { memories: rawMemories, total } = await client.listMemories(listParams);
        const memories = opts.global
          ? rawMemories.filter((m) => (m as { project_id?: string | null }).project_id === null)
          : rawMemories;
        spinner.stop();

        if (memories.length === 0) {
          info("No memories found. Create one with: remb memory add -t \"Title\" -c \"Content\"");
          return;
        }

        info(`${total} memor${total === 1 ? "y" : "ies"} found\n`);

        if (opts.format === "json") {
          console.log(JSON.stringify(memories, null, 2));
          return;
        }

        if (opts.format === "markdown") {
          for (const m of memories) {
            console.log(`### ${m.title}`);
            console.log(`- **Tier**: ${m.tier} | **Category**: ${m.category}`);
            console.log(`- **Tags**: ${m.tags.length ? m.tags.join(", ") : "none"}`);
            console.log(`- **Tokens**: ${m.token_count} | **ID**: ${m.id}`);
            console.log(`\n${m.content}\n`);
          }
          return;
        }

        // Table format
        const tierColors: Record<string, (s: string) => string> = {
          core: chalk.yellow,
          active: chalk.cyan,
          archive: chalk.dim,
        };

        for (const m of memories) {
          const tierFn = tierColors[m.tier] ?? chalk.white;
          const scope = (m as { project_id?: string | null }).project_id
            ? chalk.blue("[project]".padEnd(10))
            : chalk.magenta("[global]".padEnd(10));
          console.log(
            `${scope} ${tierFn(`[${m.tier}]`.padEnd(10))} ${chalk.bold(m.title.slice(0, 45).padEnd(47))} ${chalk.dim(m.category.padEnd(12))} ${chalk.dim(`${m.token_count}t`)}`
          );
          if (m.tags.length) {
            console.log(`                     ${chalk.dim(m.tags.map((t: string) => `#${t}`).join(" "))}`);
          }
        }
      } catch (err) {
        spinner.stop();
        handleError(err);
      }
    });
}

/* ─── remb memory update ─── */

function memoryUpdateCommand() {
  return new Command("update")
    .description("Update an existing memory")
    .argument("<id>", "Memory ID")
    .option("-t, --title <title>", "New title")
    .option("-c, --content <content>", "New content")
    .option("--tier <tier>", "New tier")
    .option("--category <category>", "New category")
    .option("--tags <tags>", "Comma-separated tags")
    .action(async (id, opts) => {
      validateUUID(id, "Memory ID");
      if (opts.title) validateStringLength(opts.title, "Title", 200);
      if (opts.content) validateContentSize(opts.content, 50);
      if (opts.tier) validateEnum(opts.tier, "tier", VALID_TIERS);
      if (opts.category) validateEnum(opts.category, "category", VALID_CATEGORIES);

      const spinner = ora("Updating memory...").start();
      try {
        const client = createApiClient();
        const params: Record<string, unknown> = {};
        if (opts.title) params.title = opts.title;
        if (opts.content) params.content = opts.content;
        if (opts.tier) params.tier = opts.tier;
        if (opts.category) params.category = opts.category;
        if (opts.tags) params.tags = opts.tags.split(",").map((t: string) => t.trim());

        const result = await client.updateMemory(id, params);
        spinner.stop();
        success(`Memory updated`);
        keyValue("Tier", result.tier);
        keyValue("Category", result.category);
        keyValue("Tokens", String(result.token_count));
      } catch (err) {
        spinner.stop();
        handleError(err);
      }
    });
}

/* ─── remb memory delete ─── */

function memoryDeleteCommand() {
  return new Command("delete")
    .alias("rm")
    .description("Delete a memory")
    .argument("<id>", "Memory ID")
    .option("-f, --force", "Skip confirmation prompt")
    .action(async (id, opts) => {
      validateUUID(id, "Memory ID");

      if (!opts.force) {
        const confirmed = await confirmAction(
          `Delete memory ${chalk.bold(id.slice(0, 8))}...?`,
        );
        if (!confirmed) {
          info("Cancelled.");
          return;
        }
      }

      const spinner = ora("Deleting memory...").start();
      try {
        const client = createApiClient();
        await client.deleteMemory(id);
        spinner.stop();
        success("Memory deleted");
      } catch (err) {
        spinner.stop();
        handleError(err);
      }
    });
}

/* ─── remb memory promote ─── */

function memoryPromoteCommand() {
  return new Command("promote")
    .description("Promote a memory to a higher tier (archive→active→core)")
    .argument("<id>", "Memory ID")
    .option("--to <tier>", "Target tier: core, active", "core")
    .action(async (id, opts) => {
      validateUUID(id, "Memory ID");
      validateEnum(opts.to, "tier", ["core", "active"]);

      const spinner = ora(`Promoting memory to ${opts.to}...`).start();
      try {
        const client = createApiClient();
        const result = await client.updateMemory(id, { tier: opts.to });
        spinner.stop();
        success(`Memory promoted to ${result.tier}`);
        keyValue("Title", result.title);
      } catch (err) {
        spinner.stop();
        handleError(err);
      }
    });
}

File: packages/cli-go/cmd/memory.go
package cmd

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"text/tabwriter"

	"github.com/useremb/remb/internal/api"
	"github.com/useremb/remb/internal/output"
	"github.com/spf13/cobra"
)

var memoryCmd = &cobra.Command{
	Use:     "memory",
	Aliases: []string{"mem"},
	Short:   "Manage AI memories — add, list, update, delete, and promote",
}

// ─── remb memory add ───────────────────────────────────────────────────

var memAddTitle string
var memAddContent string
var memAddTier string
var memAddCategory string
var memAddTags string
var memAddProject string

var memoryAddCmd = &cobra.Command{
	Use:   "add",
	Short: "Create a new memory",
	RunE:  runMemoryAdd,
}

func runMemoryAdd(cmd *cobra.Command, args []string) error {
	fmt.Print("⠋ Creating memory...")

	client, err := api.NewClient()
	if err != nil {
		fmt.Print("\r\033[K")
		output.Error(err.Error())
		os.Exit(1)
	}

	var tags []string
	if memAddTags != "" {
		for _, t := range strings.Split(memAddTags, ",") {
			t = strings.TrimSpace(t)
			if t != "" {
				tags = append(tags, t)
			}
		}
	}

	result, err := client.CreateMemory(api.CreateMemoryRequest{
		Title:       memAddTitle,
		Content:     memAddContent,
		Tier:        memAddTier,
		Category:    memAddCategory,
		Tags:        tags,
		ProjectSlug: memAddProject,
	})
	fmt.Print("\r\033[K")
	if err != nil {
		handleAPIError(err)
	}

	fmt.Println()
	output.Success("Memory created")
	output.KeyValue("ID", result.ID)
	output.KeyValue("Tier", result.Tier)
	output.KeyValue("Category", result.Category)
	output.KeyValue("Tokens", fmt.Sprintf("%d", result.TokenCount))
	return nil
}

// ─── remb memory list ──────────────────────────────────────────────────

var memListTier string
var memListCategory string
var memListSearch string
var memListProject string
var memListLimit string
var memListFormat string

var memoryListCmd = &cobra.Command{
	Use:     "list",
	Aliases: []string{"ls"},
	Short:   "List memories",
	RunE:    runMemoryList,
}

func runMemoryList(cmd *cobra.Command, args []string) error {
	fmt.Print("⠋ Fetching memories...")

	client, err := api.NewClient()
	if err != nil {
		fmt.Print("\r\033[K")
		output.Error(err.Error())
		os.Exit(1)
	}

	params := map[string]string{}
	if memListTier != "" {
		params["tier"] = memListTier
	}
	if memListCategory != "" {
		params["category"] = memListCategory
	}
	if memListSearch != "" {
		params["search"] = memListSearch
	}
	if memListProject != "" {
		params["project"] = memListProject
	}
	if memListLimit != "" {
		params["limit"] = memListLimit
	}

	result, err := client.ListMemories(params)
	fmt.Print("\r\033[K")
	if err != nil {
		handleAPIError(err)
	}

	if len(result.Memories) == 0 {
		output.Info("No memories found. Create one with: remb memory add -t \"Title\" -c \"Content\"")
		return nil
	}

	fmt.Fprintf(os.Stderr, "%sℹ%s %d memor%s found\n\n", "\033[36m", "\033[0m", result.Total, pluralize(result.Total, "y", "ies"))

	switch memListFormat {
	case "json":
		b, _ := json.MarshalIndent(result.Memories, "", "  ")
		fmt.Println(string(b))
	case "markdown":
		for _, m := range result.Memories {
			fmt.Printf("### %s\n", m.Title)
			fmt.Printf("- **Tier**: %s | **Category**: %s\n", m.Tier, m.Category)
			if len(m.Tags) > 0 {
				fmt.Printf("- **Tags**: %s\n", strings.Join(m.Tags, ", "))
			}
			fmt.Printf("- **Tokens**: %d | **ID**: %s\n", m.TokenCount, m.ID)
			fmt.Printf("\n%s\n\n", m.Content)
		}
	default:
		w := tabwriter.NewWriter(os.Stdout, 0, 0, 2, ' ', 0)
		for _, m := range result.Memories {
			tierColor := tierToColor(m.Tier)
			title := m.Title
			if len(title) > 50 {
				title = title[:47] + "..."
			}
			fmt.Fprintf(w, "%s[%s]\033[0m\t%s%s\033[0m\t%s%s\033[0m\t%s%dt\033[0m\n",
				tierColor, m.Tier, "\033[1m", title, "\033[2m", m.Category, "\033[2m", m.TokenCount)
			if len(m.Tags) > 0 {
				tagStr := make([]string, len(m.Tags))
				for i, t := range m.Tags {
					tagStr[i] = "#" + t
				}
				fmt.Fprintf(w, "\t%s%s\033[0m\t\t\n", "\033[2m", strings.Join(tagStr, " "))
			}
		}
		w.Flush()
	}

	return nil
}

// ─── remb memory update ────────────────────────────────────────────────

var memUpdateTitle string
var memUpdateContent string
var memUpdateTier string
var memUpdateCategory string
var memUpdateTags string

var memoryUpdateCmd = &cobra.Command{
	Use:   "update <id>",
	Short: "Update an existing memory",
	Args:  cobra.ExactArgs(1),
	RunE:  runMemoryUpdate,
}

func runMemoryUpdate(cmd *cobra.Command, args []string) error {
	id := args[0]
	fmt.Print("⠋ Updating memory...")

	client, err := api.NewClient()
	if err != nil {
		fmt.Print("\r\033[K")
		output.Error(err.Error())
		os.Exit(1)
	}

	req := api.UpdateMemoryRequest{}
	if memUpdateTitle != "" {
		req.Title = memUpdateTitle
	}
	if memUpdateContent != "" {
		req.Content = memUpdateContent
	}
	if memUpdateTier != "" {
		req.Tier = memUpdateTier
	}
	if memUpdateCategory != "" {
		req.Category = memUpdateCategory
	}
	if memUpdateTags != "" {
		var tags []string
		for _, t := range strings.Split(memUpdateTags, ",") {
			t = strings.TrimSpace(t)
			if t != "" {
				tags = append(tags, t)
			}
		}
		req.Tags = tags
	}

	result, err := client.UpdateMemory(id, req)
	fmt.Print("\r\033[K")
	if err != nil {
		handleAPIError(err)
	}

	fmt.Println()
	output.Success("Memory updated")
	output.KeyValue("Tier", result.Tier)
	output.KeyValue("Category", result.Category)
	output.KeyValue("Tokens", fmt.Sprintf("%d", result.TokenCount))
	return nil
}

// ─── remb memory delete ────────────────────────────────────────────────

var memoryDeleteCmd = &cobra.Command{
	Use:     "delete <id>",
	Aliases: []string{"rm"},
	Short:   "Delete a memory",
	Args:    cobra.ExactArgs(1),
	RunE:    runMemoryDelete,
}

func runMemoryDelete(cmd *cobra.Command, args []string) error {
	id := args[0]
	fmt.Print("⠋ Deleting memory...")

	client, err := api.NewClient()
	if err != nil {
		fmt.Print("\r\033[K")
		output.Error(err.Error())
		os.Exit(1)
	}

	err = client.DeleteMemory(id)
	fmt.Print("\r\033[K")
	if err != nil {
		handleAPIError(err)
	}

	fmt.Println()
	output.Success("Memory deleted")
	return nil
}

// ─── remb memory promote ───────────────────────────────────────────────

var memPromoteTo string

var memoryPromoteCmd = &cobra.Command{
	Use:   "promote <id>",
	Short: "Promote a memory to a higher tier (archive→active→core)",
	Args:  cobra.ExactArgs(1),
	RunE:  runMemoryPromote,
}

func runMemoryPromote(cmd *cobra.Command, args []string) error {
	id := args[0]
	fmt.Printf("⠋ Promoting memory to %s...", memPromoteTo)

	client, err := api.NewClient()
	if err != nil {
		fmt.Print("\r\033[K")
		output.Error(err.Error())
		os.Exit(1)
	}

	result, err := client.UpdateMemory(id, api.UpdateMemoryRequest{Tier: memPromoteTo})
	fmt.Print("\r\033[K")
	if err != nil {
		handleAPIError(err)
	}

	fmt.Println()
	output.Success(fmt.Sprintf("Memory promoted to %s", result.Tier))
	output.KeyValue("Title", result.Title)
	return nil
}

// ─── helpers ───────────────────────────────────────────────────────────

func tierToColor(tier string) string {
	switch tier {
	case "core":
		return "\033[33m" // yellow
	case "active":
		return "\033[36m" // cyan
	case "archive":
		return "\033[2m" // dim
	default:
		return ""
	}
}

func pluralize(n int, singular, plural string) string {
	if n == 1 {
		return singular
	}
	return plural
}

func init() {
	// add
	memoryAddCmd.Flags().StringVarP(&memAddTitle, "title", "t", "", "Memory title (required)")
	memoryAddCmd.Flags().StringVarP(&memAddContent, "content", "c", "", "Memory content (required)")
	memoryAddCmd.Flags().StringVar(&memAddTier, "tier", "active", "Memory tier: core, active, archive")
	memoryAddCmd.Flags().StringVar(&memAddCategory, "category", "general", "Category: preference, pattern, decision, correction, knowledge, general")
	memoryAddCmd.Flags().StringVar(&memAddTags, "tags", "", "Comma-separated tags")
	memoryAddCmd.Flags().StringVarP(&memAddProject, "project", "p", "", "Project slug")
	_ = memoryAddCmd.MarkFlagRequired("title")
	_ = memoryAddCmd.MarkFlagRequired("content")

	// list
	memoryListCmd.Flags().StringVar(&memListTier, "tier", "", "Filter by tier: core, active, archive")
	memoryListCmd.Flags().StringVar(&memListCategory, "category", "", "Filter by category")
	memoryListCmd.Flags().StringVarP(&memListSearch, "search", "s", "", "Search memories")
	memoryListCmd.Flags().StringVarP(&memListProject, "project", "p", "", "Filter by project")
	memoryListCmd.Flags().StringVarP(&memListLimit, "limit", "l", "20", "Max results")
	memoryListCmd.Flags().StringVar(&memListFormat, "format", "table", "Output format: table, json, markdown")

	// update
	memoryUpdateCmd.Flags().StringVarP(&memUpdateTitle, "title", "t", "", "New title")
	memoryUpdateCmd.Flags().StringVarP(&memUpdateContent, "content", "c", "", "New content")
	memoryUpdateCmd.Flags().StringVar(&memUpdateTier, "tier", "", "New tier")
	memoryUpdateCmd.Flags().StringVar(&memUpdateCategory, "category", "", "New category")
	memoryUpdateCmd.Flags().StringVar(&memUpdateTags, "tags", "", "Comma-separated tags")

	// promote
	memoryPromoteCmd.Flags().StringVar(&memPromoteTo, "to", "core", "Target tier: core, active")

	// register subcommands
	memoryCmd.AddCommand(memoryAddCmd)
	memoryCmd.AddCommand(memoryListCmd)
	memoryCmd.AddCommand(memoryUpdateCmd)
	memoryCmd.AddCommand(memoryDeleteCmd)
	memoryCmd.AddCommand(memoryPromoteCmd)

	// register with root
	rootCmd.AddCommand(memoryCmd)
}
